### PR TITLE
added tf.SequenceExample to ExampleGen

### PR DESCRIPTION
**Pull request by tinally**
# Add tf.Sequence Example to ExampleGen for base_example_gen and import_example_gen.

### CODE REVIEW

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 1]
```diff
@@ -88,14 +101,17 @@ def _PartitionFn(
...
+                                       bytes])
 @beam.typehints.with_output_types(beam.pvalue.PDone)
 def _WriteSplit(example_split: beam.pvalue.PCollection,
                 output_split_path: Text) -> beam.pvalue.PDone:
   """Shuffles and writes output split as serialized records in TFRecord."""
 
   def _MaybeSerialize(x):
-    return x.SerializeToString() if isinstance(x, tf.train.Example) else x
+    return x.SerializeToString() if isinstance(x, (tf.train.Example,
+      tf.train.SequenceExample)) else x
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor.py:114]: I feel this won't pass the lint

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 2]
```diff
@@ -109,14 +125,20 @@ def _MaybeSerialize(x):
...
     pipeline: beam.Pipeline,
     input_to_example: beam.PTransform,
     exec_properties: Dict[Text, Any],
     split_pattern: Text,
 ) -> beam.pvalue.PCollection:
-  """Converts input into a tf.train.Example, or a bytes (serialized proto)."""
+  """Converts input into one of the following:
+     - tf.train.Example,
+     - tf.train.SequenceExample, 
+     - a bytes (serialized proto).
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor.py:140]: I feel you might need a full doc (includes args and returns, check examples in this repo) if it's not a single line, otherwise line won't pass

File: tfx/components/example_gen/base_example_gen_executor_test.py [hunk 3]
```diff
@@ -279,10 +279,10 @@ def testInvalidFeatureBasedPartitionWithProtos(self):
...
-        RuntimeError, 'Split by `partition_feature_name` is only supported for '
-        'FORMAT_TF_EXAMPLE payload format.'):
+        RuntimeError, 'Split by `partition_feature_name` is only supported '
+        'for FORMAT_TF_EXAMPLE or FORMAT_TF_SEQUENCE_EXAMPLE payload format.'):
       example_gen.Do({}, self._output_dict, self._exec_properties)
 
 
 if __name__ == '__main__':
-  tf.test.main()
+  tf.test.main()
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor_test.py:288]: don't remove the last line (lint/format error), same apply to other files

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 4]
```diff
@@ -40,18 +40,30 @@
 DEFAULT_FILE_NAME = 'data_tfrecord'
 
 
-def _ExamplePartitionKey(record: tf.train.Example,
-                         split_config: example_gen_pb2.SplitConfig) -> bytes:
-  """Generates key for partition for tf.train.Example."""
+def _ExamplePartitionKey(
+  record: Union[tf.train.Example, tf.train.SequenceExample],
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor.py:44]: let have a unit test for both sequenceExample in base_example_gen_executor_test.py (you need another mock to create pcollection of sequenceexample, just like the one we have for now for tf.examples) and ImportExampleGen tests, we need to make sure importExampleGen can import sequenceExamples

File: tfx/components/example_gen/base_example_gen_executor.py (hunk 1 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor.py:114]: yes, you are right, this is too long to fit in 2 lines with correct formatting

File: tfx/components/example_gen/base_example_gen_executor_test.py (hunk 3 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor_test.py]: I see, I will keep this in mind

File: tfx/proto/example_gen.proto [hunk 5]
```diff
@@ -71,6 +71,9 @@ enum PayloadFormat {
 
   // Serialized tf.train.Example protocol buffer.
   FORMAT_TF_EXAMPLE = 6;
+  
+  // Serialized tf.train.SequenceExample protocol buffer.
+  FORMAT_TF_SEQUENCE_EXAMPLE = 7;
```
Comment (zhitaoli) [inline tfx/proto/example_gen.proto:76]: 7 is reserved below. Please use 12 and update line 81

Comment (tinally) [reply tfx/proto/example_gen.proto:76]: got it, i made the change

Comment (tinally) [reply tfx/proto/example_gen.proto:76]: I think internally we use 7, do they need to align? 
https://screenshot.googleplex.com/Vqc7OWfpefa

File: tfx/components/example_gen/base_example_gen_executor.py (hunk 2 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor.py]: yes, I changed this back to one line.

File: tfx/proto/example_gen.proto (hunk 5 above)
Comment (tinally) [reply tfx/proto/example_gen.proto:76]: to align with internal, i removed the reserved 7 below

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 6]
```diff
@@ -40,18 +40,33 @@
...
+    split_config: example_gen_pb2.SplitConfig
+) -> bytes:
+  """Generates key for partition."""
 
   if not split_config.HasField('partition_feature_name'):
+    if isinstance(record, bytes):
+      return record
     return record.SerializeToString(deterministic=True)
 
+  features = ""
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor.py:54]: you don't need this (and feature is not str, I guess you mean feature=None), python's scope is different, you can access the var defined in if block

Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor.py:54]: got it, thanks for the catch!

File: tfx/components/example_gen/import_example_gen/executor_test.py [hunk 7]
```diff
@@ -140,6 +140,25 @@ def testDoWithProto(self):
         self.examples.get_string_custom_property(
             utils.PAYLOAD_FORMAT_PROPERTY_NAME))
 
+  def testDoWithSequenceExamples(self):
```
Comment (1025KB) [inline tfx/components/example_gen/import_example_gen/executor_test.py:143]: create a sequence testdata

File: tfx/components/example_gen/import_example_gen/executor.py [hunk 8]
```diff
@@ -69,11 +69,13 @@ def GetInputSourceToExamplePTransform(self) -> beam.PTransform:
...
     @beam.typehints.with_input_types(beam.Pipeline)
-    @beam.typehints.with_output_types(Union[tf.train.Example, bytes])
+    @beam.typehints.with_output_types(Union[tf.train.Example,
+                                            tf.train.SequenceExample,
+                                            bytes])
     def ImportProtoOrExample(pipeline: beam.Pipeline,
                              exec_properties: Dict[Text, Any],
                              split_pattern: Text) -> beam.pvalue.PCollection:
-      """PTransform to import tf.train.Example records or serialized proto."""
+      """PTransform to import a tf.train.Example, tf.train.SequenceExample or serialized proto."""
```
Comment (1025KB) [inline tfx/components/example_gen/import_example_gen/executor.py:78]: it's a pcollection of xxx, so `s` not `a`

and you may need a full doc comment as it is not fit into a single line

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 9]
```diff
@@ -109,14 +118,16 @@ def _MaybeSerialize(x):
...
+                                        tf.train.SequenceExample,
+                                        bytes])
 def _InputToExampleOrBytes(
     pipeline: beam.Pipeline,
     input_to_example: beam.PTransform,
     exec_properties: Dict[Text, Any],
     split_pattern: Text,
 ) -> beam.pvalue.PCollection:
-  """Converts input into a tf.train.Example, or a bytes (serialized proto)."""
+  """Converts input into a tf.train.Example, tf.train.SequenceExample, or serialized proto."""
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor.py:130]: lint issue: line > 80

File: tfx/components/example_gen/import_example_gen/executor_test.py (hunk 7 above)
Comment (tinally) [reply tfx/components/example_gen/import_example_gen/executor_test.py:143]: yes, i added it under the testdata folder

File: tfx/components/example_gen/import_example_gen/executor.py (hunk 8 above)
Comment (tinally) [reply tfx/components/example_gen/import_example_gen/executor.py:78]: understand, i will change this to a full doc comment

File: tfx/components/example_gen/base_example_gen_executor.py (hunk 4 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor.py]: got it, i added a unit test for testImportSequenceExample

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 10]
```diff
@@ -109,14 +118,31 @@ def _MaybeSerialize(x):
...
     split_pattern: Text,
 ) -> beam.pvalue.PCollection:
-  """Converts input into a tf.train.Example, or a bytes (serialized proto)."""
+  """Converts input into records.
+
+  The records are tf.train.Example, tf.train.SequenceExample,
+  or serialized proto.
+
+  Args:
+    pipeline: beam pipeline.
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor.py:136]: beam -> Beam

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 11]
```diff
@@ -127,8 +153,8 @@ class BaseExampleGenExecutor(
   """Generic TFX example gen base executor.
 
   The base ExampleGen executor takes a configuration and converts external data
-  sources to TensorFlow Examples (tf.train.Example), or any other protocol
-  buffer as subclass defines.
+  sources to TensorFlow Examples (tf.train.Example), tf.train.SequenceExample,
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor.py:156]: TensorFlow Examples (tf.train.Example, tf.train.SequenceExample) or ...

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 12]
```diff
@@ -109,14 +118,31 @@ def _MaybeSerialize(x):
 
 @beam.ptransform_fn
 @beam.typehints.with_input_types(beam.Pipeline)
-@beam.typehints.with_output_types(Union[tf.train.Example, bytes])
+@beam.typehints.with_output_types(Union[tf.train.Example,
+                                        tf.train.SequenceExample,
+                                        bytes])
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor.py:123]: I wonder if we should drop _InputToExampleOrBytes, as it's basically the same as input_to_example

and could you also rename input_to_example to input_to_record

File: tfx/components/example_gen/base_example_gen_executor_test.py [hunk 13]
```diff
@@ -53,8 +53,12 @@ def _TestInputSourceToExamplePTransform(pipeline, exec_properties,
     feature['s'] = tf.train.Feature(
     ) if i % 10 == 0 and has_empty else tf.train.Feature(
         bytes_list=tf.train.BytesList(value=[tf.compat.as_bytes(str(i))]))
-    example_proto = tf.train.Example(
-        features=tf.train.Features(feature=feature))
+    if exec_properties.get('sequence_example', False):
+      example_proto = tf.train.SequenceExample(
+          context=tf.train.Features(feature=feature))
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor_test.py:58]: put sth in feature list?

File: tfx/components/example_gen/base_example_gen_executor_test.py [hunk 14]
```diff
@@ -151,6 +155,29 @@ def testDoOutputSplit(self):
...
+    example_gen.Do({}, self._output_dict, self._exec_properties)
+
+    # Check example gen outputs.
+    self.assertTrue(tf.io.gfile.exists(self._train_output_file))
+    self.assertTrue(tf.io.gfile.exists(self._eval_output_file))
+
+    # Output split ratio: train:eval=2:1.
+    self.assertGreater(
+        tf.io.gfile.GFile(self._train_output_file).size(),
+        tf.io.gfile.GFile(self._eval_output_file).size())
```
Comment (1025KB) [inline tfx/components/example_gen/base_example_gen_executor_test.py:179]: I noticed there are a lot of duplication for line 169-179 in this file, could you help us refactor it? (create a _testDo like importexamplegen executor_test)

File: tfx/components/example_gen/import_example_gen/executor.py [hunk 15]
```diff
@@ -69,11 +69,27 @@ def GetInputSourceToExamplePTransform(self) -> beam.PTransform:
...
+
+      Args:
+        pipeline: beam pipeline.
+        exec_properties: A dict of execution properties.
+          - input_base: input dir that contains tf example data.
+        split_pattern: Split.pattern in Input config, glob relative file pattern
+          that maps to input files with root directory given by input_base.
+
+      Returns:
+        PCollection of TF examples.
```
Comment (1025KB) [inline tfx/components/example_gen/import_example_gen/executor.py:91]: sth like: PCollection of record (tf.example sequence or bytes)

File: tfx/components/example_gen/import_example_gen/executor.py [hunk 16]
```diff
@@ -69,11 +69,27 @@ def GetInputSourceToExamplePTransform(self) -> beam.PTransform:
...
                              exec_properties: Dict[Text, Any],
                              split_pattern: Text) -> beam.pvalue.PCollection:
-      """PTransform to import tf.train.Example records or serialized proto."""
+      """PTransform to import records.
+
+      The records are tf.train.Example, tf.train.SequenceExample,
+      or serialized proto.
+
+      Args:
+        pipeline: beam pipeline.
```
Comment (1025KB) [inline tfx/components/example_gen/import_example_gen/executor.py:84]: beam-Beam

File: tfx/components/example_gen/import_example_gen/executor.py [hunk 17]
```diff
@@ -88,7 +104,13 @@ def ImportProtoOrExample(pipeline: beam.Pipeline,
         return (serialized_records
                 | 'ToTFExample' >> beam.Map(tf.train.Example.FromString))
 
-      raise ValueError('output_payload_format must be one of FORMAT_TF_EXAMPLE '
-                       'or FORMAT_PROTO')
+      elif (output_payload_format ==
```
Comment (1025KB) [inline tfx/components/example_gen/import_example_gen/executor.py:107]: remove the empty line between if/elif/elif

File: tfx/components/example_gen/import_example_gen/executor_test.py [hunk 18]
```diff
@@ -69,6 +69,24 @@ def check_result(got):
 
       util.assert_that(examples, check_result)
 
+  def testImportSequenceExample(self):
```
Comment (1025KB) [inline tfx/components/example_gen/import_example_gen/executor_test.py:72]: we feel don't need this test, as it's mainly for testing _ImportSerializedRecord (which doesn't contain a branch in it)

File: tfx/components/example_gen/import_example_gen/executor_test.py [hunk 19]
```diff
@@ -140,6 +158,32 @@ def testDoWithProto(self):
...
+        
+    exec_properties = {
+        utils.INPUT_BASE_KEY:
+            self._input_data_dir,
+        utils.INPUT_CONFIG_KEY:
+            self._input_config,
+        utils.OUTPUT_CONFIG_KEY:
+            self._output_config,
+        utils.OUTPUT_DATA_FORMAT_KEY:
+            example_gen_pb2.PayloadFormat.FORMAT_TF_SEQUENCE_EXAMPLE,
```
Comment (1025KB) [inline tfx/components/example_gen/import_example_gen/executor_test.py:177]: just confirm, if change to format tf example, will it fail?

File: tfx/components/testdata/external/README.md [hunk 20]
```diff
@@ -13,3 +13,42 @@ df.to_parquet('data.parquet', index=False)
 
 Avro data file can be generated by running the script in
 `avro/generateAvroFile.py`
+
```
Comment (1025KB) [inline tfx/components/testdata/external/README.md:16]: line 1 "Note that the testing files here are same as [taxi data]" is not true after adding this sequence file, update it

File: tfx/components/testdata/external/README.md [hunk 21]
```diff
@@ -13,3 +13,42 @@ df.to_parquet('data.parquet', index=False)
...
+
+A TF Sequence Example is created under tfrecord_sequence folder.
```
Comment (1025KB) [inline tfx/components/testdata/external/README.md:18]: tfrecord_sequence folder contains a tfrecord file of dummy tf.train.SequenceExamples

The format of the testdata is as follows.

**(use code format)**
``` 
context: {
  feature: {
    key  : "i"
    value: {...}
  }
  feature: {
    key  : "f"
    value: {...}
  }
}
feature_lists: {
  feature_list: {
    key  : "s"
    value: {
      feature: {...}
      feature: {...}
    }
  }
}
```

File: tfx/components/example_gen/import_example_gen/executor_test.py (hunk 19 above)
Comment (tinally) [reply tfx/components/example_gen/import_example_gen/executor_test.py:177]: no it doesn't fail.

but the same thing applies to testImportExample -- if I change format to tf sequence example, it doesn't fail.

File: tfx/components/example_gen/base_example_gen_executor.py (hunk 12 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor.py:123]: should i just delete _InputToExampleOrBytes? or do we keep it for now?

File: tfx/components/example_gen/base_example_gen_executor_test.py (hunk 14 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor_test.py:179]: yes i see, i added _testDo

File: tfx/components/example_gen/base_example_gen_executor.py (hunk 12 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor.py]: i deleted _InputToExampleOrBytes for now

File: tfx/components/testdata/external/README.md [hunk 22]
```diff
@@ -1,5 +1,4 @@
-Note that the testing files here are same as [taxi data](https://github.com/tensorflow/tfx/blob/master/tfx/examples/chicago_taxi_pipeline/data/simple/data.csv)
-but different formats.
+The testing files here contain [taxi data](https://github.com/tensorflow/tfx/blob/master/tfx/examples/chicago_taxi_pipeline/data/simple/data.csv) with different formats and a tf.train.SequenceExample file.
```
Comment (1025KB) [inline tfx/components/testdata/external/README.md:1]: # Taxi Data
....


# Dummy SequenceExample
...

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 23]
```diff
@@ -268,7 +262,7 @@ def Do(
       input_dict: Input dict from input key to a list of Artifacts. Depends on
         detailed example gen implementation.
       output_dict: Output dict from output key to a list of Artifacts.
-        - examples: splits of tf examples.
+        - examples: splits of TF records.
```
Comment (hughmiao) [inline tfx/components/example_gen/base_example_gen_executor.py:265]: please update the rest of the docstring (e.g., in Do()
--the output is indented to be serialized tf.train.Examples protocol buffer )

File: tfx/components/example_gen/import_example_gen/executor.py [hunk 24]
```diff
@@ -41,18 +41,18 @@ def _ImportSerializedRecord(  # pylint: disable=invalid-name
   Note that each input split will be transformed by this function separately.
 
   Args:
-    pipeline: beam pipeline.
+    pipeline: Beam pipeline.
     exec_properties: A dict of execution properties.
-      - input_base: input dir that contains tf example data.
+      - input_base: input dir that contains TF record data.
```
Comment (hughmiao) [inline tfx/components/example_gen/import_example_gen/executor.py:46]: note, TFRecord is a keyword for a file format, 
consider to simply use `contains records` or `contains input data` etc. 
please revisit the usage of this term  (`TF record`, `TFRecord`) in the rest of the change.

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 25]
```diff
@@ -176,15 +172,15 @@ def GenerateExamplesByBeam(
       pipeline: beam.Pipeline,
       exec_properties: Dict[Text, Any],
   ) -> Dict[Text, beam.pvalue.PCollection]:
-    """Converts input source to TF example splits based on configs.
+    """Converts input source to TF record splits based on configs.
```
Comment (hughmiao) [inline tfx/components/example_gen/base_example_gen_executor.py:175]: for the usage for `TF record`, please see the comment below.

File: tfx/components/example_gen/base_example_gen_executor_test.py [hunk 26]
```diff
@@ -124,58 +130,46 @@ def testDoInputSplit(self):
...
 
-    # Output split ratio: train:eval=2:1.
-    self.assertGreater(
-        tf.io.gfile.GFile(self._train_output_file).size(),
-        tf.io.gfile.GFile(self._eval_output_file).size())
+    self._testDo(self._exec_properties)
 
   def testFeatureBasedPartition(self):
-    # Add output config to exec proterties.
+    # Update output config in exec proterties.
```
Comment (hughmiao) [inline tfx/components/example_gen/base_example_gen_executor_test.py:172]: shall we add a test to test feature based partition with sequence example?

Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor_test.py:172]: yes sure, i added a test for this

File: tfx/components/example_gen/base_example_gen_executor.py (hunk 25 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor.py]: i replace TF record with 'serialized record', is this ok?

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 27]
```diff
@@ -131,12 +131,12 @@ class BaseExampleGenExecutor(
 
   The conversion is done in `GenerateExamplesByBeam` as a Beam pipeline, which
   validates the configuration, reads the external data sources, converts the
-  record in the input source to TF record if needed, and splits the examples if
-  the output split config is given. Then the executor's `Do` writes the results
-  in splits to the output path.
+  record in the input source to tf.Example or tf.SequenceExample if needed,
```
Comment (hughmiao) [inline tfx/components/example_gen/base_example_gen_executor.py:134]: nit, given we also allow subclass to support any bytes, consider to reword this to  `input source to any supported output payload formats (e.g., tf.Example or tf.SequenceExample)`

File: tfx/components/example_gen/base_example_gen_executor.py [hunk 28]
```diff
@@ -131,12 +131,12 @@ class BaseExampleGenExecutor(
...
+  and splits the examples if the output split config is given.
+  Then the executor's `Do` writes the results in splits to the output path.
 
   For simple custom ExampleGens, the details of transforming input data
-  record(s) to TF record(s) is expected to be given in
+  record(s) to tf.Example or tf.SequenceExample is expected to be given in
```
Comment (hughmiao) [inline tfx/components/example_gen/base_example_gen_executor.py:139]: ditto, to specific output formats

File: tfx/components/example_gen/base_example_gen_executor.py (hunk 27 above)
Comment (tinally) [reply tfx/components/example_gen/base_example_gen_executor.py]: got it, this makes sense

**Comment by 1025KB** [comment]
manually closed this one as this [PR](https://github.com/tensorflow/tfx/pull/2083) is merged to master
